ratis-server/src/main/java/org/apache/ratis/statemachine/impl/SimpleStateMachineStorage.java

```diff
@@ -109,16 +118,28 @@ public void cleanupOldSnapshots(SnapshotRetentionPolicy snapshotRetentionPolicy)
 
     if (allSnapshotFiles.size() > snapshotRetentionPolicy.getNumSnapshotsRetained()) {
       allSnapshotFiles.sort(Comparator.comparing(SingleFileSnapshotInfo::getIndex).reversed());
-      List<File> snapshotFilesToBeCleaned = allSnapshotFiles.subList(
-              snapshotRetentionPolicy.getNumSnapshotsRetained(), allSnapshotFiles.size()).stream()
-          .map(singleFileSnapshotInfo -> singleFileSnapshotInfo.getFile().getPath().toFile())
-          .collect(Collectors.toList());
-      for (File snapshotFile : snapshotFilesToBeCleaned) {
-        LOG.info("Deleting old snapshot at {}", snapshotFile.getAbsolutePath());
-        final boolean deleted = FileUtils.deleteFileQuietly(snapshotFile);
-        if (deleted) {
-          final File md5file = MD5FileUtil.getDigestFileForFile(snapshotFile);
-          FileUtils.deleteFileQuietly(md5file);
+      allSnapshotFiles.subList(numSnapshotsRetained, allSnapshotFiles.size())
+          .stream()
+          .map(SingleFileSnapshotInfo::getFile)
+          .map(FileInfo::getPath)
+          .forEach(snapshotPath -> {
+            LOG.info("Deleting old snapshot at {}", snapshotPath.toAbsolutePath());
+            FileUtils.deletePathQuietly(snapshotPath);
+          });
+      // clean up the md5 files if the corresponding snapshot file does not exist
+      try (DirectoryStream<Path> stream = Files.newDirectoryStream(stateMachineDir.toPath(),
+          SNAPSHOT_MD5_FILTER)) {
+        for (Path md5path : stream) {
+          Path md5FileNamePath = md5path.getFileName();
+          if (md5FileNamePath == null) {
+            continue;
+          }
```

Comment:
FYI, `md5FileNamePath` cannot be null anymore since `SNAPSHOT_MD5_FILTER` won't return it.

Reply:
This is to satisfy the findbugs.